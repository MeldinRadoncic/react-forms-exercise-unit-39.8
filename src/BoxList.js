import React,{useState} from "react";
import Box from "./Box";
import NewBoxForm from "./NewBoxForm";
import {v4 as uuid} from 'uuid'




const BoxList = () => {

    const [boxes,setBoxes] = useState([])



    const addBox = (newBoxes) => {
        setBoxes(data => [...data,newBoxes])
        
      }

      const removeBox = (id) => {
        setBoxes(boxes => boxes.filter(
         box => box.id !== id
        ))
      }

      const newBox = boxes.map(box => (
        <Box
          key={box.id}
          id={box.id}
          width={box.width}
          height={box.height}
          color={box.color}
          handleRemove={removeBox}
        />
      ));

    return (
        <div>
            <h1>Color Box Maker</h1>
            <NewBoxForm addBox={addBox} />
           {newBox}


        </div>
    )
}

export default BoxList;