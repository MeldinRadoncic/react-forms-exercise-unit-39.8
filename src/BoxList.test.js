import {fireEvent, render, screen } from '@testing-library/react';

import React from "react";
import BoxList from "./BoxList";


it('Renders without crashing', () => {
    render(<BoxList/>)
})


it('matches snapshot', () => {
    const {asFragment} = render(<BoxList/>)
    expect (asFragment()).toMatchSnapshot();
} )


it('Display Alert message if inputs are empty', function() {

     render(<BoxList/>)
     const alertMock = jest.spyOn(window,'alert').mockImplementation(); 

    // Get inputs by label text
  const color =screen.getByLabelText('Color')
  const width = screen.getByLabelText('Width')
  const height = screen.getByLabelText('Height')

  const addBoxBtn = screen.getByText('Add Box')
    expect (screen.queryByText('Color:red')).not.toBeInTheDocument();
    
    // If one or more inputs are empty display alert after submitting form
    fireEvent.change(width, {target:{value:'100'}});
    fireEvent.change(height, {target:{value:''}});

    fireEvent.change(color, {target:{value:'red'}});

    fireEvent.click(addBoxBtn);
    expect(alertMock).toHaveBeenCalledTimes(1)

})

it('Add new Bpx', function() {

    render(<BoxList/>)

   // Get inputs by label text
 const color =screen.getByLabelText('Color')
 const width = screen.getByLabelText('Width')
 const height = screen.getByLabelText('Height')

 const addBoxBtn = screen.getByText('Add Box')
   
   fireEvent.change(width, {target:{value:'100'}});
   fireEvent.change(height, {target:{value:'100'}});
   fireEvent.change(color, {target:{value:'red'}});

   fireEvent.click(addBoxBtn);

   const newBox = screen.getByTestId('box')
   const removeBtn = screen.getByTestId('removeBtn')


   expect(newBox).toBeInTheDocument();
   expect(removeBtn).toBeInTheDocument();
})

it('Removes a box', () => {
    render(<BoxList/>)

    const color = screen.getByLabelText('Color')
    const width = screen.getByLabelText('Width')
    const height = screen.getByLabelText('Height')

    const addBox = screen.getByText('Add Box')

    fireEvent.change(color, {target:{value:'red'}})
    fireEvent.change(width, {target:{value:'300'}})
    fireEvent.change(height, {target:{value:'300'}})

    fireEvent.click(addBox)
    const newBox = screen.getByTestId('box')
    expect(newBox).toBeInTheDocument();

    const removeBtn = screen.getByTestId('removeBtn')


    fireEvent.click(removeBtn)
    expect(newBox).not.toBeInTheDocument();


})