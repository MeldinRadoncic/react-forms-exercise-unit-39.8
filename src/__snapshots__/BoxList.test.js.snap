// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Color Box Maker
    </h1>
    <div>
      <form>
        <label
          for="color"
        >
          Color
        </label>
        <input
          id="color"
          name="color"
          placeholder="Enter a color..."
          type="text"
          value=""
        />
        <label
          for="width"
        >
          Width
        </label>
        <input
          id="width"
          name="width"
          placeholder="Enter a width..."
          type="text"
          value=""
        />
        <label
          for="height"
        >
          Height
        </label>
        <input
          id="height"
          name="height"
          placeholder="Enter a height..."
          type="text"
          value=""
        />
        <button>
          Add Box
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
